rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+the new shapes which incurs additional overhead. The goal of this RFC is to propose
+a solution to reduce that overhead.
+
+## Background
+
+There is a reason why the oneDNN API requires users to specify tensors dimensions
+to create a primitive. The library may use the information to generate a kernel
+specifically for the given shapes to get the best performance.
+
+However, the information may not always be used by the library meaning that the
+generated kernels can be used for different shapes, which makes requirement to
+always specify tensors dimensions is overly strict.
+
+There was an attempt in the past to relax the requirement by introducing so-called
+runtime dimensions support. The idea was to allow users specify a placeholder
+(`DNNL_RUNTIME_DIM_VAL`) instead of the actual dimensions. The general rule for
+the feature was that the more actual dimensions users provide the better
+performance they can get.
+
+It's obvious that in order to support runtime dimensions the kernels must not
+depend on the shapes. Or there should be different kernels that are specific to
+the amount of information provided. And the kernels should also demonstrate
+reasonably good performance otherwise, the creation overhead will be replaced by
+a suboptimal performance overhead, which will defeat the purpose of having the
+runtime dimensions feature.
+
+The runtime dimensions feature is not a silver bullet, this is why the RFC will
+consider another option to reduce the primitive creation overhead.
+
+## Proposal
+
+There are two options to reduce primitive creation overhead:
+1. Extend the runtime dimensions feature for the relevant primitives
+2. Introduce a cache for kernels
+
+### Option 1: Cache for Kernels (recommended)
+
+The idea is to introduce a cache for kernels (hereinafter *cache*). The cache will
+be global similar to the primitive cache.
+
+Pros:
+* Doesn't require users to modify their code to make the feature work
+* Eliminates the kernel creation overhead completely when it's possible
+* Doesn't require to expose kernel implementation details. For example, for
+the runtime dimensions feature we will have to document what dimensions can
+be runtime without introducing performance regressions
+
+Cons:
+* Primitive and primitive descriptor creation overhead will stay
+
+#### Key and Value
+
+The cache implementation will reside in the `common` part of the library. In order
+to abstract away `key` and `value` entities we will need to introduce the
+following abstractions:
+
+```cpp
+namespace kernel_cache {
+// Key
+struct key_t {
+    key_t(key_impl_t *impl);
+    virtual ~key_t() = default;
+
+    bool operator==(const key_t &other) const;
+    size_t hash() const;
+private:
+    std::shared_ptr<key_impl_t> impl_;
+};
+
+// The `key_impl_t` class will be the base class for all `key` entities.
+// Each engine kind and vendor will be able to define it the way the need.
```

Comment:
Will this be inherited only once per engine/vendor?
I would assume that there is a need for different key/comparison function for at least each primitive that uses it (e.g. matmul, conv), and also per implementation (e.g. brgemm convolution has different kernel descriptor than "legacy" jitted kernels).

Do we have an idea of how many derived classes there will be ?

Reply:
This is something that the backend teams will have to define. This can be done in different ways. This is possible to have just one key and one value classes, this can be done by introducing some primitive specific configuration structures that will be used in a union in the key along with other relevant data. The value can be just a vector of `kernel_t`, for example.

But this is on the backend teams and ideally they would need to define it through the RFC process. 
I believe @rjoursler has already implemented something.

Reply:
My expectation is that there will be one key implementation per implementation in the library. A rough estimate is 1 per final derived primitive descriptor. While there are opportunities to reuse these keys between implementations, it creates a significant vector for passing uninitialized data into hashing/comparison. This exact issue crop up multiple times in `cl_cache`. As such, such sharing needs validate with tooling such as Memory Sanitizer or requiring the class is `constexpr` constructible. My personal opinion is that the cost/benefit tradeoff of sharing `key_impl_t` is not enough warrant its use in general.